src/Microsoft.Health.Fhir.Tests.Common/ICustomFhirClient.cs

```diff
@@ -0,0 +1,154 @@
+using Microsoft.Health.Fhir.Tests.Common.FixtureParameters;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Tests.Common
+{
+    public interface ICustomFhirClient
+    {
+        Format Format { get; }
+
+        (bool SecurityEnabled, string AuthorizeUrl, string TokenUrl) SecuritySettings { get; }
+
+        HttpClient HttpClient { get; }
+
+        Task RunAsUser(TestUser user, TestApplication clientApplication);
+
+        Task RunAsClientApplication(TestApplication clientApplication);
+
+        Task<FhirResponse<T>> CreateAsync<T>(T resource)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> CreateAsync<T>(string uri, T resource)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> ReadAsync<T>(string resourceType, string resourceId)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> ReadAsync<T>(string uri)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> VReadAsync<T>(string resourceType, string resourceId, string versionId)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> UpdateAsync<T>(T resource, string ifMatchVersion = null)
+            where T : ResourceElement;
+
+        Task<FhirResponse<T>> UpdateAsync<T>(string uri, T resource, string ifMatchVersion = null)
+            where T : ResourceElement;
+
+        Task<FhirResponse> DeleteAsync<T>(T resource)
+            where T : ResourceElement;
+
+        Task<FhirResponse> DeleteAsync(string uri);
+
+        Task<FhirResponse> HardDeleteAsync<T>(T resource)
+            where T : ResourceElement;
+
+        Task<FhirResponse<ResourceElement>> SearchAsync(string resourceType, string query = null, int? count = null);
+
+        Task<FhirResponse<ResourceElement>> SearchAsync(string url);
+
+        Task<FhirResponse<ResourceElement>> SearchPostAsync(string resourceType, params (string key, string value)[] body);
+
+        Task<string> ExportAsync(Dictionary<string, string> queryParams);
+
+        ////StringContent CreateStringContent(Resource resource);
+
+        Task EnsureSuccessStatusCodeAsync(HttpResponseMessage response);
+
+        Task<FhirResponse<T>> CreateResourceElementResponseAsync<T>(HttpResponseMessage response)
+            where T : ResourceElement;
+
+        ////Task<FhirResponse<T>> CreateResponseAsync<T>(HttpResponseMessage response)
+        ////    where T : Resource;
+
+        Task SetupAuthenticationAsync(TestApplication clientApplication, TestUser user = null);
+
+        Task<string> GetBearerToken(TestApplication clientApplication, TestUser user);
+
+        List<KeyValuePair<string, string>> GetAppSecuritySettings(TestApplication clientApplication);
+
+        List<KeyValuePair<string, string>> GetUserSecuritySettings(TestApplication clientApplication, TestUser user);
+
+        Task GetSecuritySettings(string fhirServerMetadataUrl);
+
+        ResourceElement GetDefaultObservation();
```

Comment:
Extract these from FhirClient to some sort of resource element extensions/provider.